File: 0037-sudoku-solver/0037-sudoku-solver.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        int fillCount = 0;
        bool printed = false;

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    fillCount++;
                }
            }
        }

        solve(fillCount, board);

        // print the soduku solution
        // for (int i = 0; i < 9; i++) {
        //     for (int j = 0; j < 9; j++) {
        //         cout << board[i][j] << "\t";
        //     }
        //     cout << endl;
        // }
        
    }

private:
       bool solve(int fillCount, vector<vector<char>>& board) {
        if (fillCount == 81) {
            return true;
        }

        int emptyRow, emptyCol;
        bool exitLoop = false;
        for (int i = 0; i < 9 && !exitLoop; i++) {
            for (int j = 0; j < 9 && !exitLoop; j++) {
                if (board[i][j] == '.') {
                    emptyRow = i;
                    emptyCol = j;
                    exitLoop = true;
                }
            }
        }

        bool solved = false;
        for (int k = 1; k <= 9 && !solved; k++) {
            char ch = '0' + k;
            if (canBePlaced(ch, emptyRow, emptyCol, board)) {
                board[emptyRow][emptyCol] = ch;
                solved = solve(fillCount + 1, board);
                if (solved) {
                    return true;
                }
                board[emptyRow][emptyCol] = '.';
            }
        }

        return false;
    }

    bool canBePlaced(char k, int i, int j, vector<vector<char>>& board) {
        // check same row
        for (int c = 0; c < 9; c++) {
            if (board[i][c] == k) {
                return false;
            }
        }

        // check same column
        for (int r = 0; r < 9; r++) {
            if (board[r][j] == k) {
                return false;
            }
        }

        // check in its own box
        pair<int, int> rowBoundary = getStartAndEnd(i);
        pair<int, int> colBoundary = getStartAndEnd(j);
        for (int r = rowBoundary.first; r <= rowBoundary.second; r++) {
            for (int c = colBoundary.first; c <= colBoundary.second; c++) {
                if (board[r][c] == k) {
                    return false;
                }
            }
        }

        // if all the above conditions are false then return true
        return true;
    }

    pair<int, int> getStartAndEnd(int n) {
        pair<int, int> res;
        if (0 <= n && n <= 2) {
            res.first = 0;
            res.second = 2;
        } else if (3 <= n && n <= 5) {
            res.first = 3;
            res.second = 5;
        } else {
            res.first = 6;
            res.second = 8;
        }

        return res;
    }
};

File: 0037-sudoku-solver/0037-sudoku-solver.java
class Solution {
    /**
     * The main method that initiates the Sudoku solving process.
     * The input board is modified in-place to contain the solution.
     *
     * @param board The 9x9 grid representing the Sudoku puzzle, with '.' for empty cells.
     */
    public void solveSudoku(char[][] board) {
        // Start the recursive solving process.
        solve(board);
    }

    /**
     * The recursive backtracking helper method.
     * It tries to fill the board and returns true if a solution is found.
     *
     * @param board The Sudoku grid.
     * @return true if the board is solved, false otherwise.
     */
    private boolean solve(char[][] board) {
        // Iterate through each cell of the board.
        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                // Find an empty cell represented by '.'.
                if (board[row][col] == '.') {
                    // Try placing numbers '1' through '9' in the empty cell.
                    for (char num = '1'; num <= '9'; num++) {
                        // Check if the current number placement is valid.
                        if (canBePlaced(board, row, col, num)) {
                            // If valid, place the number.
                            board[row][col] = num;

                            // Recursively call solve() for the next empty cell.
                            // If the recursive call returns true, a solution was found.
                            if (solve(board)) {
                                return true;
                            } else {
                                // If not, backtrack: undo the move and try the next number.
                                board[row][col] = '.';
                            }
                        }
                    }
                    // If no number ('1'-'9') works for this cell, trigger backtracking.
                    return false;
                }
            }
        }
        // If the entire board is filled (no empty cells found), a solution is found.
        return true;
    }

    /**
     * Checks if a given number can be placed in a specific cell without
     * violating Sudoku rules.
     *
     * @param board The Sudoku grid.
     * @param row The row index of the cell.
     * @param col The column index of the cell.
     * @param num The character ('1'-'9') to check.
     * @return true if the placement is valid, false otherwise.
     */
    private boolean canBePlaced(char[][] board, int row, int col, char num) {
        // Check for the same number in the same row and column.
        for (int i = 0; i < 9; i++) {
            if (board[row][i] == num) {
                return false; // Check row
            }
            if (board[i][col] == num) {
                return false; // Check column
            }
        }

        // Check for the same number in the 3x3 sub-box.
        int boxStartRow = (row / 3) * 3;
        int boxStartCol = (col / 3) * 3;

        for (int r = boxStartRow; r < boxStartRow + 3; r++) {
            for (int c = boxStartCol; c < boxStartCol + 3; c++) {
                if (board[r][c] == num) {
                    return false;
                }
            }
        }

        // If all checks pass, the number can be placed.
        return true;
    }
}